web/src/server/api/routers/comments.ts

```diff
@@ -288,4 +261,104 @@ export const commentsRouter = createTRPCRouter({
         });
       }
     }),
+  getTraceCommentCountsBySessionId: protectedProjectProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        sessionId: z.string(),
+        queryClickhouse: z.boolean().default(false),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      try {
+        return await measureAndReturnApi({
+          input,
+          operation: "comments.getTraceCommentCountsBySessionId",
+          user: ctx.session.user ?? undefined,
+          pgExecution: async () => {
+            const session = await ctx.prisma.traceSession.findFirst({
+              where: {
+                id: input.sessionId,
+                projectId: input.projectId,
+              },
+              include: {
+                traces: {
+                  select: {
+                    id: true,
+                  },
+                },
+              },
+            });
+            if (!session) {
+              throw new TRPCError({
+                code: "NOT_FOUND",
+                message: "Session not found in project",
+              });
+            }
+
+            const allTraceCommentCounts = await ctx.prisma.$queryRaw<
+              Array<{ objectId: string; count: bigint }>
+            >`
+            SELECT object_id as "objectId", COUNT(*) as count
+            FROM comments
+            WHERE project_id = ${input.projectId}
+            AND object_type = 'TRACE'
+            GROUP BY object_id
+          `;
+
+            const traceIds = new Set(
+              session.traces.map((t: { id: string }) => t.id),
+            );
+            return new Map(
+              allTraceCommentCounts
+                .filter((c) => traceIds.has(c.objectId))
+                .map(({ objectId, count }) => [objectId, Number(count)]),
+            );
+          },
+          clickhouseExecution: async () => {
+            const postgresSession = await ctx.prisma.traceSession.findFirst({
+              where: {
+                id: input.sessionId,
+                projectId: input.projectId,
+              },
+            });
+
+            if (!postgresSession) {
+              throw new TRPCError({
+                code: "NOT_FOUND",
+                message: "Session not found in project",
+              });
+            }
+
+            const clickhouseTraces = await getTracesIdentifierForSession(
+              input.projectId,
+              input.sessionId,
+            );
+
```

Comment:
actually, we can improve the above. No need to check for postgres tracesession at all. You can just query clickhouse for all traces in the sessionId.

Reply:
😍 true, will remove redundant code!